File: python/bob/bob.py
def response(hey_bob: str) -> str:
    """
    return what Bob will reply to someone when they say something to him or ask him a question.
    """

    if len(bob := hey_bob.strip()) == 0:
        return "Fine. Be that way!"
    
    return (
        "Whatever.",
        "Sure.",
        "Whoa, chill out!",
        "Calm down, I know what I'm doing!"
    )[is_yell(bob) + is_question(bob)]


def is_yell(text: str) -> bool:
    return 2 if text.isupper() else 0


def is_question(text: str) -> bool:
    return 1 if text.endswith("?") else 0

File: java/bob/src/main/java/Bob.java
import java.util.function.Predicate;
import java.util.regex.Pattern;

final class Bob {
  private static final Predicate<String> isQuestion =
      Pattern.compile(".*\\?\\s*", Pattern.DOTALL).asMatchPredicate();
  private static final Predicate<String> hasUppercase =
      Pattern.compile("\\p{Upper}", Pattern.UNICODE_CHARACTER_CLASS).asPredicate();
  private static final Predicate<String> hasLowercase =
      Pattern.compile("\\p{Lower}", Pattern.UNICODE_CHARACTER_CLASS).asPredicate();
  private static final Predicate<String> isYell = hasUppercase.and(hasLowercase.negate());
  private static final Predicate<String> isYellQuestion = isQuestion.and(isYell);

  String hey(String input) {
    if (input.isBlank()) return "Fine. Be that way!";
    if (isYellQuestion.test(input)) return "Calm down, I know what I'm doing!";
    if (isYell.test(input)) return "Whoa, chill out!";
    if (isQuestion.test(input)) return "Sure.";
    return "Whatever.";
  }
}
